lib/Dialect/FIRRTL/FIRRTLConnectionGraph.cpp

```diff
@@ -0,0 +1,222 @@
+//===- FIRRTLConnectionGraph.h - Graph of connections in FIRRTL -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides utilities for working with FIRRTL operations using LLVM
```

Comment:
What kind of graph is the result?

What are its properties, what guarantees does it provide, how does one think about it?
How should this be maintained such that users will continue to work without deeply understanding their use cases and assumptions?

What does the graph for a simple module look like?  What happens when there are regions?

Please clarify.  Ideally adding tests and documentation.

At a glance this code provides adapters to walk operations and their users, and transparently look through destination of a connection operation.  Is that about right?
(Implementation aside, is this what it's aiming to be, conceptually?)

What makes it a "connection" graph?

Reply:
Okay, the GraphTraits are where things are clicked for me, was wondering how any of these fit together.

Nodes are Operation*'s.
They have children (edges to):
* Module -> Each Operation within its body block (I may be missing something where it's not top-level only?)
* ConnectionLike  -> DestinationOperand
* Other Operation's -> Users of all results (auto-skipping connection-dest uses).

Module is structural container, the outermost "Graph" at least for now.  Maybe all "structural" operations should work this way? Not sure, anyway.
Operations within are the nodes.

Edges are provided for use->user, skipping connection destinations (we want those edges to go the other way).
For a connection op, it has an edge to its destination.

Could we add the bits needed for GraphWriter so we can generate a GraphViz (DOT) graph for a simple module, possibly in documentation or something?

So there's the structural aspect (Module -> Operations, no connectivity on those "edges" necessarily), but otherwise this looks like its a .. "data-flow" graph kinda?

-----

As discussed, this adapter strategy has challenges re:indexing, a common challenge in many of our analyses/passes.

It does not capture dataflow through symbols (not sure it should just saying), so users should be careful to keep that in mind.

Reply:
You're grokking it. It's a graph describing the dataflow. I usually think of what this is doing as each iterator needs to define all "children" ops that it needs to visit---or it needs to iterate over everything which it drives.

As you state, this works fine for anything which is a subfield/subindex/subaccess. There are XFAIL tests for this in the follow-on PR here that show missed opportunities for sinking operations into layers: https://github.com/llvm/circt/pull/6492/files#diff-02a186dd2d7d7235891c5204bf87aeee4b561bcc62393b5cbe60c00a84bb4dbe

The problem (arguably with FIRRTL dialect) is that, when subfield/subindex/subaccess are involved, an operation iterator cannot make a local decision about its dataflow children without doing additional analysis. Additionally, some operations may be both children and parents (e.g., if the same subfield is used as a source and sink).

There are a couple of ways that this may be alleviated: 

1. FIRRTL Dialect is changed to represent results as read and write as opposed to read-or-write-or-both depending on use.
2. These graph traits are changed to iterate over declarations. This requires more complex analysis to properly iterate and may be computationally expensive to do so.
3. ???

The main question here is this enough of a foundation to start from or must this work for subfield/subindex/subaccess before merging be figured out?